src/hotspot/share/classfile/classLoaderExt.cpp

```diff
@@ -131,6 +138,34 @@ void ClassLoaderExt::setup_module_paths(JavaThread* current) {
   process_module_table(current, met);
 }
 
+bool ClassLoaderExt::has_jar_suffix(const char* filename) {
+  const char* dot = strrchr(filename, '.');
+  if (dot != nullptr && strcmp(dot + 1, "jar") == 0) {
+    return true;
+  }
+  return false;
+}
+
+void ClassLoaderExt::extract_jar_files_from_path(const char* path, GrowableArray<const char*>* module_paths) {
+  DIR* dirp = os::opendir(path);
+  if (dirp == nullptr && errno == ENOTDIR && has_jar_suffix(path)) {
+    module_paths->append(path);
+  } else {
+    struct dirent* dentry;
+    while ((dentry = os::readdir(dirp)) != nullptr) {
+      const char* file_name = dentry->d_name;
+      if (has_jar_suffix(file_name)) {
+        size_t full_name_len = strlen(path) + strlen(file_name) + strlen(os::file_separator()) + 1;
+        char* full_name = NEW_RESOURCE_ARRAY(char, full_name_len);
+        int n = os::snprintf(full_name, full_name_len, "%s%s%s", path, os::file_separator(), file_name);
+        assert((size_t)n == full_name_len - 1, "Unexpected number of characters in string");
+        module_paths->append(full_name);
```

Comment:
Can this case be handled: --module-path=dir

- Dump time :  dir contains only mod1.jar
- Run time :  dir contains only mod1.jar and mod2.jmod

Reply:
It should work because the jmod file won't be added to the `module_paths`.

Reply:
In my scenario, will the FMG be used? If so, the program won't be able to load the code in mod2.jmod, so the behavior will be wrong. Could you add a test case for this?

Reply:
It turns out that jmod cannot be used during runtime.
From `jdk.intenal.module.ModulePath`:
```
            // not recognized
            String msg;
            if (!isLinkPhase && entry.toString().endsWith(".jmod")) {
                msg = "JMOD format not supported at execution time";
            } else {
                msg = "Module format not recognized";
            }
```
I added some tests with jmod.